installdir/htmerge.1
.TH htmerge 1 "21 July 1997"
.\" NAME should be all caps, SECTION should be 1-8, maybe w/ subsection
.\" other parms are allowed: see man(7), man(1)
.SH NAME
htmerge \- create document index and word database for the ht://Dig
search engine
.SH SYNOPSIS
.B htmerge
.I "[options]"
.SH "DESCRIPTION"
.I Htmerge
is used to create a document index and word database
from the files that were created by \fIhtdig\fR. These databases are
then used by
.I htsearch
to perform the actual searched. 
.SH OPTIONS
.TP
.B \-a
Use alternate work files. Tells htdig to append
.I .work
to database files, causing a second copy of the database to be
built. This allows the original files to be used by htsearch
during the indexing run. 
.TP
.B \-c \fIconfigfile\fR
Use the specified
.I configfile
instead of the default. 
.TP
.B \-d
Prevent the document index from being created. 
.TP
.B \-s
Print statistics about the document and word databases
after htmerge has finished. 
.TP
.B \-v
Run in verbose mode. This will provide some hints as to the
progress of the merge. This can be useful when running
htmerge interactively since some parts (especially the word
database creation) can take a very long time. 
.TP
.B \-w
Prevent the word database from being created. 
.SH ENVIRONMENT
.TP
.B TMPDIR
In addition to the command line options, the environment
variable
.I TMPDIR
will be used to designate the directory where
intermediate files are stored during the sorting process. 
.SH "FILES"
.TP
.B /etc/htdig/htdig.conf
The default configuration file.
.SH "SEE ALSO"
Please refer to the HTML pages (in the htdig-doc package)
.B /usr/share/doc/htdig-doc/html/index.html
and the manual pages
.I htdig(1)
and 
.I htsearch(1)
for a detailed description of ht://Dig and its commands.
.SH AUTHOR
This manual page was written by Christian Schwarz, modified
by Stijn de Bekker, based on the HTML documentation of ht://Dig.


ChangeLog
Wed Feb 23 16:53:40 2000  Gilles Detillieux  <grdetil@scrc.umanitoba.ca>

	* htcommon/defaults.cc (backlink_factor, external_parser,
	local_default_doc, local_urls, local_urls_only, local_user_urls):
	Add some updates from 3.1.5's attrs.html.
	* htdoc/attrs.html, htdoc/cf_byname.html, htdoc/cf_byprog.html:
	Regenerate using cf_generate.pl.

Wed Feb 23 15:11:51 2000  Gilles Detillieux  <grdetil@scrc.umanitoba.ca>

	[ Improve htsearch's HTML 4.0 compliance ]
	* htsearch/TemplateList.cc (createFromString): Use file name rather
	than internal name to select builtin-* templates, use $&(TITLE) and
	$&(URL) in templates and quote HTML tag parameters.
	* installdir/long.html, installdir/short.html: Use $&(TITLE) and
	$&(URL) in templates and quote HTML tag parameters.
	* htsearch/Display.cc (setVariables): quote all HTML tag parameters
	in generated select lists.
	* installdir/footer.html, installdir/header.html,
	installdir/nomatch.html, installdir/search.html,
	installdir/syntax.html, installdir/wrapper.html:
	Use $&(var) where appropriate, and quote HTML tag parameters.
	* installdir/htdig.conf: quote all HTML tag parameters.

Wed Feb 23 13:40:27 2000  Gilles Detillieux  <grdetil@scrc.umanitoba.ca>

	* htcommon/URL.h (encodeURL): Change list of valid characters to
	include only unreserved ones.
	* htcommon/cgi.cc (init): Allow "&" and ";" as input param. separators.
	* htsearch/Display.cc (createURL): Encode each parameter separately,
	using new unreserved list, before piecing together query string, to
	allow characters like "?=&" within parameters to be encoded.

Wed Feb 23 13:22:29 2000  Gilles Detillieux  <grdetil@scrc.umanitoba.ca>

	* htlib/URL.cc (ServerAlias): Fix server_aliases processing to prevent
	infinite loop (as for local_urls in PR#688).

Wed Feb 23 12:49:52 2000  Gilles Detillieux  <grdetil@scrc.umanitoba.ca>

	* htlib/HtDateTime.h, htlib/HtDateTime.cc: change Httimegm() method
	to HtTimeGM(), to avoid conflict with Httimegm() C function, so we
	don't need "::" override, for Mac OS X.
	* htlib/htString.h, htlib/String.cc: change write() method to
	Write(), to avoid conflict with write() function, so we don't need
	"::" override, for Mac OS X.

Wed Feb 23 12:17:46 2000  Gilles Detillieux  <grdetil@scrc.umanitoba.ca>

	* htlib/Configuration.cc(Read): Fixed to allow final line without
	terminating newline character, rather than ignoring it.

Wed Feb 23 12:01:01 2000  Gilles Detillieux  <grdetil@scrc.umanitoba.ca>

	* htdig/Retriever.cc (GetLocal, GetLocalUser): Add URL-decoding
	enhancements to local_urls, local_default_urls & local_default_doc,
	to allow hex encoding of special characters.

Wed Feb 23 19:14:29 2000  Loic Dachary  <loic@ceic.com>

 	* htcommon/conf_parser.cxx: regenerated from conf_parser.yxx

Wed Feb 23 19:04:16 2000  Loic Dachary  <loic@ceic.com>

	* test/test_functions.in: inconditionaly remove existing test/var
	  directory before runing tests to prevent accidents.

	* htcommon/URL.cc (URL): fixed String->char warning

	* htcommon/defaults.cc (wordlist_compress): defaults to true

Tue Feb 22 17:09:10 2000  Gilles Detillieux  <grdetil@scrc.umanitoba.ca>

	* htdig/HTML.cc(parse, do_tag): Fix handling of <img alt=...> text
	and parsing of words in meta tags, to to proper word separation.
	* htlib/HtWordType.h, htlib/HtWordType.cc: Add HtWordToken() function,
	to replace strtok() in HTML parser.

Tue Feb 22 16:21:25 2000  Gilles Detillieux  <grdetil@scrc.umanitoba.ca>

	* htlib/URL.cc (ctor, normalizePath): Fix PR#779, to handle relative
	URLs correctly when there's a trailing ".." or leading "//".

Tue Feb 22 14:09:26 2000  Gilles Detillieux  <grdetil@scrc.umanitoba.ca>

	* htdig/Document.cc (RetrieveLocal): Handle common extensions for
	text/plain, application/pdf & application/postscript.

Mon Feb 21 17:25:21 2000  Gilles Detillieux  <grdetil@scrc.umanitoba.ca>

	* contrib/htdig-3.2.0.spec: Fixed %post script to add more
	descriptive entries in htdig.conf, made cron script a config file,
	updated to 3.2.0b2.

	* contrib/conv_doc.pl, contrib/parse_doc.pl: Added comments to show
	Warren Jones's updates in change history.

Mon Feb 21 17:09:13 2000  Gilles Detillieux  <grdetil@scrc.umanitoba.ca>

	* htcommon/HtConfiguration.h, htcommon/conf_parser.yxx,
	htlib/Configuration.h, htlib/Configuration.cc: split Add() method
	into Add() and AddParsed(), so that only config attributes get parsed.
	Use AddParsed() only in Read() and Defaults().

Sat Feb 19 04:56:47 2000  Geoff Hutchison  <ghutchis@wso.williams.edu>

	* htnet/Connection.h, htnet/Connection.cc: Renamed methods with
	capitals to remove the need to use ::-escaped library calls.

Fri Feb 18 14:40:50 2000  Loic Dachary  <loic@ceic.com>

	* test/conf/access.conf.in: removed cookies. Not used and some
	  httpd are not compiled with usertrack.

Wed Feb 16 12:15:08 2000  Vadim Chekan  <vadim@etc.lviv.ua>

	* htcommon/Makefile.am replaced conf.tab.cc.h by conf_parser.h in
	noinst_HEADERS

	* htcommon/conf_parser.yxx,conf_parser.lxx,HtConfiguration.cc,
	HtConfiguration.h: added copyright and Id:

	* htcommon/cgi.cc(init): fixed bug: array must be free by 
	delete [] buf, not just delete buf;

Tue Feb 15 23:16:14 2000  Geoff Hutchison  <ghutchis@wso.williams.edu>

	* htnet/HtHTTP.cc (isParsable): Remove application/pdf as a
	default type--it is now handled through the ExternalParser
	interface if at all.

	* htcommon/defaults.cc: Remove pdf_parser attribute.

	* htdoc/attrs.html, htdoc/cf_byname.html, htdoc/cf_byprog.html:
	Regenerate using cf_generate.pl.

	* htdig/Document.cc (getParsable): Remove PDF once and for all
	(hopefully).

	* htdig/ExternalParser.cc (parse): Ditto.

	* configure.in: Remove check for PDF_PARSER.

	* configure: Regenerate using autoconf

	* htdig/Makefile.am: Remove PDF.cc and PDF.h.

	* Makefile.in, */Makefile.in: Regenerate using automake --foreign

Tue Feb 15 12:02:39 EET 2000  Vadim Chekan  <vadim@etc.lviv.ua>

	* htcommon/HtConfiguration.cc,HtConfiguration.h: fixed bug discovered

httools/htmerge.cc
//
// htmerge.cc
//
// htmerge: Merges two databases and/or updates databases to remove 
//          old documents and ensures the databases are consistent.
//          Calls db.cc, docs.cc, and/or words.cc as necessary
//
// Part of the ht://Dig package   <http://www.htdig.org/>
// Copyright (c) 1999-2000 The ht://Dig Group
// For copyright details, see the file COPYING in your distribution
// or the GNU General Public License version 2 or later
// <http://www.gnu.org/copyleft/gpl.html>
//
// $Id: htmerge.cc,v 1.1.2.3 2000/08/09 18:06:52 ghutchis Exp $
//

#ifdef HAVE_CONFIG_H
#include "htconfig.h"
#endif /* HAVE_CONFIG_H */

#include "WordContext.h"
#include "good_strtok.h"
#include "defaults.h"
#include "DocumentDB.h"
#include "HtURLCodec.h"
#include "HtWordList.h"
#include "HtWordReference.h"
#include "htString.h"

#include <fstream.h>
#include <stdio.h>
#include <unistd.h>
#include <stdlib.h>
#include <ctype.h>
#include <string.h>

// If we have this, we probably want it.
#ifdef HAVE_GETOPT_H
#include <getopt.h>
#endif

class CallbackData : public Object
{
public:
  CallbackData(HtWordList * w, Dictionary * d, int o)
        { word_db = w; dup_ids = d; docIDOffset = o; }

  HtWordList	*word_db;
  Dictionary	*dup_ids;
  int		docIDOffset;
};

int
OverrideCallback(WordList * wl,
                 WordDBCursor &,
                 const WordReference * w,
                 Object & d)
{
  CallbackData		&data = ((CallbackData &)d);
  HtWordReference	*ht_wr = (HtWordReference *)w;
  String		docIDKey;

  docIDKey << ht_wr->DocID();
  if (!((data.dup_ids)->Exists(docIDKey)))
    {
      ht_wr->DocID(ht_wr->DocID() + data.docIDOffset);
      (data.word_db)->Override(*ht_wr);
    }
    
  return OK;
}

int
DeleteCallback(WordList * wl,
               WordDBCursor &,
               const WordReference * w,
               Object & d)
{
  CallbackData		&data = ((CallbackData &)d);
  HtWordReference	*ht_wr = (HtWordReference *)w;
  String		docIDKey;

  docIDKey << ht_wr->DocID();
  if ((data.dup_ids)->Exists(docIDKey))
    (data.word_db)->Delete(*ht_wr);
    
  return OK;
}


// This config is used for merging multiple databses
HtConfiguration    merge_config;

int		verbose = 0;
int		stats = 0;

// Component procedures
void mergeDB();
void usage();
void reportError(char *msg);

//*****************************************************************************
// int main(int ac, char **av)
//
int main(int ac, char **av)
{
    int			alt_work_area = 0;
    String		configfile = DEFAULT_CONFIG_FILE;
    String              merge_configfile = 0;
    int			c;
    extern char		*optarg;

    while ((c = getopt(ac, av, "svm:c:dwa")) != -1)
    {
	switch (c)
	{
	    case 'd':
		break;
	    case 'w':
		break;
	    case 'c':
		configfile = optarg;
		break;
	    case 'm':
	      	merge_configfile = optarg;
	      	break;
	    case 'v':
		verbose++;
		break;
	    case 's':
		break;
	    case 'a':
		alt_work_area++;
		break;
	    case '?':
		usage();
		break;
	}
    }

    config.Defaults(&defaults[0]);

    if (access((char*)configfile, R_OK) < 0)
    {
	reportError(form("Unable to find configuration file '%s'",
			 configfile.get()));
    }
	
    config.Read(configfile);

    //
    // Check url_part_aliases and common_url_parts for
    // errors.
    String url_part_errors = HtURLCodec::instance()->ErrMsg();

    if (url_part_errors.length() != 0)
      reportError(form("Invalid url_part_aliases or common_url_parts: %s",
                       url_part_errors.get()));

    if (merge_configfile.length())
    {
    	merge_config.Defaults(&defaults[0]);
	if (access((char*)merge_configfile, R_OK) < 0)
    	{
	reportError(form("Unable to find configuration file '%s'",
			 merge_configfile.get()));
    	}
	merge_config.Read(merge_configfile);
    }

    if (alt_work_area != 0)
    {
	String	configValue;

	configValue = config["word_db"];
	if (configValue.length() != 0)
	{
	    configValue << ".work";
	    config.Add("word_db", configValue);
	}

	configValue = config["doc_db"];
	if (configValue.length() != 0)
	{
	    configValue << ".work";
	    config.Add("doc_db", configValue);
	}

	configValue = config["doc_index"];
	if (configValue.length() != 0)
	{
	    configValue << ".work";
	    config.Add("doc_index", configValue);
	}

	configValue = config["doc_excerpt"];
	if (configValue.length() != 0)
	{
	    configValue << ".work";
	    config.Add("doc_excerpt", configValue);
	}
    }

    WordContext::Initialize(config);

    if (merge_configfile.length())
    {
	// Merge the databases specified in merge_configfile into the current
	// databases. Do this first then update the other databases as usual
	// Note: We don't have to specify anything, it's all in the config vars

	mergeDB();
    }

    return 0;
}

//*****************************************************************************
// void mergeDB()
//
void
mergeDB()
{
    DocumentDB	merge_db, db;
    List	*urls;
    Dictionary  merge_dup_ids, db_dup_ids; // Lists of DocIds to ignore
    int         docIDOffset;

    const String doc_index = config["doc_index"];
    if (access(doc_index, R_OK) < 0)
    {
	reportError(form("Unable to open document index '%s'", (const char*)doc_index));
    }
    const String doc_excerpt = config["doc_excerpt"];
    if (access(doc_excerpt, R_OK) < 0)
    {
	reportError(form("Unable to open document excerpts '%s'", (const char*)doc_excerpt));
    }
    const String doc_db = config["doc_db"];    
    if (db.Open(doc_db, doc_index, doc_excerpt) < 0)
    {
	reportError(form("Unable to open/create document database '%s'",
			 (const char*)doc_db));
    }


    const String merge_doc_index = merge_config["doc_index"];    
    if (access(merge_doc_index, R_OK) < 0)
    {
	reportError(form("Unable to open document index '%s'", (const char*)merge_doc_index));
    }
    const String merge_doc_excerpt = merge_config["doc_excerpt"];    
    if (access(merge_doc_excerpt, R_OK) < 0)
    {
	reportError(form("Unable to open document excerpts '%s'", (const char*)merge_doc_excerpt));
    }
    const String merge_doc_db = merge_config["doc_db"];
    if (merge_db.Open(merge_doc_db, merge_doc_index, merge_doc_excerpt) < 0)
    {
	reportError(form("Unable to open document database '%s'",
			 (const char*)merge_doc_db));
    }

    // Start the merging by going through all the URLs that are in
    // the database to be merged
        
    urls = merge_db.URLs();
    // This ensures that every document added from merge_db has a unique ID
    // in the new database
    docIDOffset = db.NextDocID();

    urls->Start_Get();
    String		*url;
    String		id;
    while ((url = (String *) urls->Get_Next()))
    {
	DocumentRef	*ref = merge_db[url->get()];
	DocumentRef     *old_ref = db[url->get()];
	if (!ref)
	    continue;

	if (old_ref)
	  {
	    // Oh well, we knew this would happen. Let's get the duplicate
	    // And we'll only use the most recent date.

	    if ( old_ref->DocTime() >= ref->DocTime() )
	      {
		// Cool, the ref we're merging is too old, just ignore it
		char        str[20];
		sprintf(str, "%d", ref->DocID());
		merge_dup_ids.Add(str, 0);
		
		if (verbose > 1)
		  {
		    cout << "htmerge: Duplicate, URL: " << url << " ignoring merging copy   \n";
		    cout.flush();
		  }
	      }
	    else
	      {
		// The ref we're merging is newer, delete the old one and add
		char        str[20];
		sprintf(str, "%d", old_ref->DocID());
		db_dup_ids.Add(str, 0);
		db.Delete(old_ref->DocID());
		ref->DocID(ref->DocID() + docIDOffset);
		db.Add(*ref);
                if (verbose > 1)
                  {
                    cout << "htmerge: Duplicate, URL: ";
		    cout << url->get() << " ignoring destination copy   \n";
                    cout.flush();
                  }
	      }
	  }
	else
	  {
	    // It's a new URL, just add it, making sure to load the excerpt
	    merge_db.ReadExcerpt(*ref);
	    ref->DocID(ref->DocID() + docIDOffset);
	    db.Add(*ref);
	    if (verbose > 1)
	      {
		cout << "htmerge: Merged URL: " << url->get() << "    \n";
		cout.flush();
	      }
	  }
        delete ref;
	delete old_ref;
    }    
    delete urls;
    
    // As reported by Roman Dimov, we must update db.NextDocID()
    // because of all the added records...
    db.IncNextDocID( merge_db.NextDocID() );
    merge_db.Close();
    db.Close();

    // OK, after merging the doc DBs, we do the same for the words

    HtWordList	mergeWordDB(config), wordDB(config);

    if (wordDB.Open(config["word_db"], O_RDWR) < 0)
    {
	reportError(form("Unable to open/create document database '%s'",
			 (const char*)config["word_db"]));
    }

    if (mergeWordDB.Open(merge_config["word_db"], O_RDONLY) < 0)
    {
	reportError(form("Unable to open document database '%s'",
			 (const char *)merge_config["word_db"]));
    }

    // Start the merging by going through all the URLs that are in
    // the database to be merged
        
    CallbackData data(&wordDB, &merge_dup_ids, docIDOffset);
    WordSearchDescription description(OverrideCallback, (Object *)&data);
    mergeWordDB.Walk(description);

    CallbackData data(&wordDB, &db_dup_ids, 0);
    WordSearchDescription description(DeleteCallback, (Object *)&data);
    wordDB.Walk(description);

    // Cleanup--just close the two word databases
    mergeWordDB.Close();
    wordDB.Close();
}


//*****************************************************************************
// void usage()
//   Display program usage information
//
void usage()
{
    cout << "usage: htmerge [-v][-c configfile][-m merge_configfile]\n";
    cout << "This program is part of ht://Dig " << VERSION << "\n\n";
    cout << "Options:\n";
    cout << "\t-v\tVerbose mode.  This increases the verbosity of the\n";
    cout << "\t\tprogram.  Using more than 2 is probably only useful\n";
    cout << "\t\tfor debugging purposes.  The default verbose mode\n";
    cout << "\t\tgives a progress on what it is doing and where it is.\n\n";
    cout << "\t-m merge_configfile\n";
    cout << "\t\tMerge the databases specified into the databases specified\n";
    cout << "\t\tby -c or the default.\n\n";
    cout << "\t-c configfile\n";
    cout << "\t\tUse the specified configuration file instead on the\n";
    cout << "\t\tdefault.\n\n";
    cout << "\t-a\tUse alternate work files.\n";
    cout << "\t\tTells htmerge to append .work to database files causing\n";
    cout << "\t\ta second copy of the database to be built.  This allows\n";
    cout << "\t\toriginal files to be used by htsearch during the indexing\n";
    cout << "\t\trun.\n\n";
    exit(0);
}


//*****************************************************************************
// Report an error and die
//
void reportError(char *msg)
{
    cout << "htmerge: " << msg << "\n\n";
    exit(1);
}
